File: laravel-backend/app/Services/RiskService.php
<?php

namespace App\Services;

use App\DataSources\AlphaVantage;
use App\DataSources\YahooFinance;
use Illuminate\Support\Facades\Log;

/**
 * Portfolio-level risk computations.
 */
class RiskService
{
    private YahooFinance $yahooFinance;
    private AlphaVantage $alphaVantage;

    public function __construct(YahooFinance $yahooFinance, AlphaVantage $alphaVantage)
    {
        $this->yahooFinance = $yahooFinance;
        $this->alphaVantage = $alphaVantage;
    }

    public function computePortfolioRisk(array $holdings, array $quotes): array
    {
        $weights = $this->computePositionWeights($holdings, $quotes);
        $tickers = array_map(fn($h) => $h['ticker'], $holdings);

        return [
            'portfolio_beta' => $this->computePortfolioBeta($weights, $tickers),
            'sector_concentration' => $this->computeSectorConcentration($weights, $quotes),
            'position_sizing' => $this->computePositionSizingAlerts($weights),
            'correlation' => $this->computeCorrelationMatrix($tickers),
            'drawdowns' => $this->computeDrawdownAnalysis($tickers, $quotes),
            'stop_loss_alerts' => $this->computeStopLossAlerts($tickers, $quotes),
        ];
    }

    private function computePositionWeights(array $holdings, array $quotes): array
    {
        $positions = [];
        $total = 0;

        foreach ($holdings as $h) {
            $ticker = $h['ticker'];
            $price = $quotes[$ticker]['current_price'] ?? null;
            $shares = $h['shares'];
            $value = $price ? $shares * $price : 0;
            $total += $value;
            $positions[$ticker] = ['value' => $value, 'shares' => $shares];
        }

        foreach ($positions as $ticker => &$pos) {
            $pos['weight_pct'] = $total > 0 ? round($pos['value'] / $total * 100, 2) : 0;
        }

        return $positions;
    }

    private function computePortfolioBeta(array $weights, array $tickers): array
    {
        $betas = [];
        foreach ($tickers as $t) {
            try {
                $info = $this->yahooFinance->fetchInfoSafe($t);
                $beta = $info['beta'] ?? null;
                $betas[$t] = $beta ? round((float) $beta, 2) : 1.0;
            } catch (\Throwable $e) {
                $betas[$t] = 1.0;
            }
        }

        $totalWeight = array_sum(array_column($weights, 'weight_pct'));
        if ($totalWeight == 0) {
            return ['value' => 1.0, 'interpretation' => 'No position data', 'individual' => $betas];
        }

        $portfolioBeta = 0;
        foreach ($tickers as $t) {
            $portfolioBeta += ($betas[$t] ?? 1.0) * ($weights[$t]['weight_pct'] ?? 0) / $totalWeight;
        }
        $portfolioBeta = round($portfolioBeta, 2);

        if ($portfolioBeta > 1.2) {
            $pct = (int) (($portfolioBeta - 1) * 100);
            $interp = "Aggressive — portfolio moves ~{$pct}% more than the market";
        } elseif ($portfolioBeta < 0.8) {
            $pct = (int) ((1 - $portfolioBeta) * 100);
            $interp = "Defensive — portfolio moves ~{$pct}% less than the market";
        } else {
            $interp = 'Moderate — portfolio roughly tracks the market';
        }

        return ['value' => $portfolioBeta, 'interpretation' => $interp, 'individual' => $betas];
    }

    private function computeSectorConcentration(array $weights, array $quotes): array
    {
        $sectors = [];
        foreach ($weights as $ticker => $w) {
            $sector = $quotes[$ticker]['sector'] ?? 'Unknown';
            $sectors[$sector] = ($sectors[$sector] ?? 0) + $w['weight_pct'];
        }

        arsort($sectors);
        $sectors = array_map(fn($v) => round($v, 1), $sectors);

        $warnings = [];
        foreach ($sectors as $sector => $pct) {
            if ($pct > 40) {
                $warnings[] = "{$sector} at {$pct}% — high concentration risk";
            }
        }

        return ['sectors' => $sectors, 'warnings' => $warnings];
    }

    private function computePositionSizingAlerts(array $weights, float $threshold = 20.0): array
    {
        $alerts = [];
        $maxPos = ['ticker' => null, 'weight_pct' => 0];

        foreach ($weights as $ticker => $w) {
            if ($w['weight_pct'] > $maxPos['weight_pct']) {
                $maxPos = ['ticker' => $ticker, 'weight_pct' => $w['weight_pct']];
            }
            if ($w['weight_pct'] > $threshold) {
                $alerts[] = [
                    'ticker' => $ticker,
                    'weight_pct' => $w['weight_pct'],
                    'alert' => "Position exceeds {$threshold}% threshold",
                ];
            }
        }

        return [
            'alerts' => $alerts,
            'max_position' => $maxPos,
            'position_count' => count($weights),
        ];
    }

    private function computeCorrelationMatrix(array $tickers, int $periodDays = 130): array
    {
        $tickers = array_slice($tickers, 0, 10);
        if (count($tickers) < 2) {
            return ['high_pairs' => [], 'avg_correlation' => 0];
        }

        try {
            $closesDict = [];
            foreach ($tickers as $t) {
                $history = $this->alphaVantage->fetchHistory($t, $periodDays);
                if (! empty($history)) {
                    $closesDict[$t] = array_column($history, 'close');
                }
            }

            if (count($closesDict) < 2) {
                return ['high_pairs' => [], 'avg_correlation' => 0];
            }

            // Compute returns
            $returns = [];
            foreach ($closesDict as $t => $closes) {
                $ret = [];
                for ($i = 1; $i < count($closes); $i++) {
                    $ret[] = $closes[$i - 1] > 0 ? ($closes[$i] - $closes[$i - 1]) / $closes[$i - 1] : 0;
                }
                $returns[$t] = $ret;
            }

            // Compute correlation for each pair
            $tickerKeys = array_keys($returns);
            $highPairs = [];
            $allCorrs = [];

            for ($i = 0; $i < count($tickerKeys); $i++) {
                for ($j = $i + 1; $j < count($tickerKeys); $j++) {
                    $t1 = $tickerKeys[$i];
                    $t2 = $tickerKeys[$j];
                    $corr = $this->correlation($returns[$t1], $returns[$t2]);
                    $allCorrs[] = $corr;

                    if (abs($corr) > 0.8) {
                        $highPairs[] = [
                            'pair' => [$t1, $t2],
                            'correlation' => round($corr, 3),
                            'risk' => 'High co-movement — limited diversification benefit',
                        ];
                    }
                }
            }

            $avg = ! empty($allCorrs) ? round(array_sum($allCorrs) / count($allCorrs), 3) : 0;

            return ['high_pairs' => $highPairs, 'avg_correlation' => $avg];
        } catch (\Throwable $e) {
            return ['high_pairs' => [], 'avg_correlation' => 0];
        }
    }

    private function computeDrawdownAnalysis(array $tickers, array $quotes): array
    {
        $drawdowns = [];
        $worst = ['ticker' => null, 'drawdown_pct' => 0];

        foreach ($tickers as $t) {
            $q = $quotes[$t] ?? [];
            $high = $q['fifty_two_week_high'] ?? null;
            $current = $q['current_price'] ?? null;

            if ($high && $current && $high > 0) {
                $dd = round((($current - $high) / $high) * 100, 2);
                $drawdowns[$t] = ['drawdown_pct' => $dd, 'from_high' => $high, 'current' => $current];
                if ($dd < ($worst['drawdown_pct'] ?? 0)) {
                    $worst = ['ticker' => $t, 'drawdown_pct' => $dd];
                }
            } else {
                $drawdowns[$t] = ['drawdown_pct' => null, 'from_high' => $high, 'current' => $current];
            }
        }

        $drawdowns['_worst'] = $worst;
        return $drawdowns;
    }

    private function computeStopLossAlerts(array $tickers, array $quotes, float $defaultStopPct = 0.15): array
    {
        $alerts = [];
        foreach ($tickers as $t) {
            $q = $quotes[$t] ?? [];
            $high = $q['fifty_two_week_high'] ?? null;
            $current = $q['current_price'] ?? null;
            if (! $high || ! $current) {
                continue;
            }

            $stopLevel = round($high * (1 - $defaultStopPct), 2);
            if ($current < $stopLevel) {
                $alerts[] = [
                    'ticker' => $t,
                    'current' => $current,
                    'stop_level' => $stopLevel,
                    'from_high' => $high,
                    'status' => 'BELOW_STOP_LOSS',
                ];
            } elseif ($current < $stopLevel * 1.05) {
                $alerts[] = [
                    'ticker' => $t,
                    'current' => $current,
                    'stop_level' => $stopLevel,
                    'from_high' => $high,
                    'status' => 'NEAR_STOP_LOSS',
                ];
            }
        }
        return $alerts;
    }

    private function correlation(array $x, array $y): float
    {
        $n = min(count($x), count($y));
        if ($n < 2) return 0;

        $x = array_slice($x, 0, $n);
        $y = array_slice($y, 0, $n);

        $meanX = array_sum($x) / $n;
        $meanY = array_sum($y) / $n;

        $cov = 0;
        $varX = 0;
        $varY = 0;

        for ($i = 0; $i < $n; $i++) {
            $dx = $x[$i] - $meanX;
            $dy = $y[$i] - $meanY;
            $cov += $dx * $dy;
            $varX += $dx * $dx;
            $varY += $dy * $dy;
        }

        $denom = sqrt($varX * $varY);
        return $denom > 0 ? $cov / $denom : 0;
    }
}

File: backend/app/services/risk_service.py
"""Portfolio-level risk computations."""

import logging
from datetime import datetime, timedelta

import numpy as np
import pandas as pd
from app.data_sources.massive import fetch_info, fetch_history_days
from app.data_sources.fred import fetch_risk_free_rate

logger = logging.getLogger(__name__)


def compute_portfolio_risk(holdings: list[dict], quotes: dict) -> dict:
    """Compute comprehensive portfolio risk metrics."""
    weights = _compute_position_weights(holdings, quotes)
    tickers = [h["ticker"] for h in holdings]

    beta = _compute_portfolio_beta(weights, tickers)
    sectors = _compute_sector_concentration(weights, quotes)
    sizing = _compute_position_sizing_alerts(weights)
    correlation = _compute_correlation_matrix(tickers)
    drawdowns = _compute_drawdown_analysis(tickers, quotes)
    stop_loss = _compute_stop_loss_alerts(tickers, quotes)

    return {
        "portfolio_beta": beta,
        "sector_concentration": sectors,
        "position_sizing": sizing,
        "correlation": correlation,
        "drawdowns": drawdowns,
        "stop_loss_alerts": stop_loss,
    }


def compute_ticker_risk(ticker: str) -> dict:
    """Compute per-ticker risk metrics: alpha, beta, Sharpe ratio, Monte Carlo."""
    ticker_history = fetch_history_days(ticker, 365)
    spy_history = fetch_history_days("SPY", 365)

    if len(ticker_history) < 30 or len(spy_history) < 30:
        return {"error": f"Insufficient history for {ticker}"}

    # Align by date
    spy_map = {h["date"]: h["close"] for h in spy_history}
    aligned_dates = []
    aligned_ticker_prices = []
    aligned_spy_prices = []
    for h in ticker_history:
        if h["date"] in spy_map:
            aligned_dates.append(h["date"])
            aligned_ticker_prices.append(h["close"])
            aligned_spy_prices.append(spy_map[h["date"]])

    if len(aligned_ticker_prices) < 30:
        return {"error": f"Insufficient aligned data for {ticker}"}

    t_prices = np.array(aligned_ticker_prices)
    s_prices = np.array(aligned_spy_prices)

    # Daily returns
    t_returns = np.diff(t_prices) / t_prices[:-1]
    s_returns = np.diff(s_prices) / s_prices[:-1]

    # Risk-free rate
    rf = fetch_risk_free_rate()
    if rf is None:
        rf = 0.043
    daily_rf = rf / 252

    # Beta
    cov_matrix = np.cov(t_returns, s_returns)
    beta = float(cov_matrix[0][1] / cov_matrix[1][1]) if cov_matrix[1][1] != 0 else 1.0

    # Annualized return & volatility
    ticker_ann_return = float(np.mean(t_returns) * 252)
    market_ann_return = float(np.mean(s_returns) * 252)
    ticker_ann_std = float(np.std(t_returns, ddof=1) * np.sqrt(252))

    # Jensen's Alpha
    alpha = ticker_ann_return - (rf + beta * (market_ann_return - rf))

    # Sharpe Ratio
    sharpe = (ticker_ann_return - rf) / ticker_ann_std if ticker_ann_std != 0 else 0.0

    # Monte Carlo Simulation (GBM)
    num_sims = 1000
    num_days = 126  # ~6 months
    last_price = float(t_prices[-1])
    mu = ticker_ann_return
    sigma = ticker_ann_std
    dt = 1 / 252

    rng = np.random.default_rng(42)
    Z = rng.standard_normal((num_sims, num_days))
    daily_log_returns = (mu - 0.5 * sigma**2) * dt + sigma * np.sqrt(dt) * Z
    cumulative = np.cumsum(daily_log_returns, axis=1)
    price_paths = last_price * np.exp(cumulative)

    start_col = np.full((num_sims, 1), last_price)
    price_paths = np.hstack([start_col, price_paths])

    # Percentile bands
    bands = {}
    for p in [5, 25, 50, 75, 95]:
        bands[f"p{p}"] = [round(v, 2) for v in np.percentile(price_paths, p, axis=0).tolist()]

    # Projection dates (skip weekends)
    last_date = datetime.strptime(aligned_dates[-1], "%Y-%m-%d")
    projection_dates = []
    current = last_date
    for _ in range(num_days + 1):
        projection_dates.append(current.strftime("%Y-%m-%d"))
        current += timedelta(days=1)
        while current.weekday() >= 5:
            current += timedelta(days=1)

    # Interpretations
    beta_interp = (
        "High volatility vs market" if beta > 1.3
        else "Low volatility vs market" if beta < 0.7
        else "Roughly tracks the market"
    )
    alpha_interp = (
        "Outperforming risk-adjusted expectations" if alpha > 0.02
        else "Underperforming risk-adjusted expectations" if alpha < -0.02
        else "Performing near risk-adjusted expectations"
    )
    sharpe_interp = (
        "Excellent risk-adjusted return" if sharpe > 1.0
        else "Good risk-adjusted return" if sharpe > 0.5
        else "Moderate risk-adjusted return" if sharpe > 0
        else "Poor risk-adjusted return"
    )

    return {
        "ticker": ticker,
        "period": "1y",
        "risk_free_rate": round(rf, 4),
        "beta": {"value": round(beta, 3), "interpretation": beta_interp},
        "alpha": {
            "value": round(alpha, 4),
            "annualized_pct": round(alpha * 100, 2),
            "interpretation": alpha_interp,
        },
        "sharpe_ratio": {"value": round(sharpe, 3), "interpretation": sharpe_interp},
        "annualized_return": round(ticker_ann_return * 100, 2),
        "annualized_volatility": round(ticker_ann_std * 100, 2),
        "monte_carlo": {
            "dates": projection_dates,
            "bands": bands,
            "num_simulations": num_sims,
            "num_days": num_days,
            "start_price": round(last_price, 2),
        },
    }


def _compute_position_weights(holdings: list[dict], quotes: dict) -> dict:
    """Compute $ value and % weight for each position."""
    positions = {}
    total = 0

    for h in holdings:
        ticker = h["ticker"]
        price = quotes.get(ticker, {}).get("current_price")
        shares = h["shares"]
        value = shares * price if price else 0
        total += value
        positions[ticker] = {"value": value, "shares": shares}

    for ticker in positions:
        positions[ticker]["weight_pct"] = (
            round(positions[ticker]["value"] / total * 100, 2) if total > 0 else 0
        )

    return positions


def _compute_portfolio_beta(weights: dict, tickers: list[str]) -> dict:
    """Weighted-average beta."""
    betas = {}
    for t in tickers:
        try:
            info = fetch_info(t)
            beta = info.get("beta")
            betas[t] = round(beta, 2) if beta else 1.0
        except Exception:
            betas[t] = 1.0

    total_weight = sum(w["weight_pct"] for w in weights.values())
    if total_weight == 0:
        return {"value": 1.0, "interpretation": "No position data", "individual": betas}

    portfolio_beta = sum(
        betas.get(t, 1.0) * weights[t]["weight_pct"] / total_weight for t in tickers
    )
    portfolio_beta = round(portfolio_beta, 2)

    if portfolio_beta > 1.2:
        interp = f"Aggressive — portfolio moves ~{int((portfolio_beta - 1) * 100)}% more than the market"
    elif portfolio_beta < 0.8:
        interp = f"Defensive — portfolio moves ~{int((1 - portfolio_beta) * 100)}% less than the market"
    else:
        interp = "Moderate — portfolio roughly tracks the market"

    return {"value": portfolio_beta, "interpretation": interp, "individual": betas}


def _compute_sector_concentration(weights: dict, quotes: dict) -> dict:
    """Sector allocation with warnings."""
    sectors = {}
    for ticker, w in weights.items():
        sector = quotes.get(ticker, {}).get("sector", "Unknown")
        sectors[sector] = sectors.get(sector, 0) + w["weight_pct"]

    sectors = {k: round(v, 1) for k, v in sorted(sectors.items(), key=lambda x: -x[1])}
    warnings = [
        f"{sector} at {pct}% — high concentration risk"
        for sector, pct in sectors.items()
        if pct > 40
    ]

    return {"sectors": sectors, "warnings": warnings}


def _compute_position_sizing_alerts(weights: dict, threshold: float = 20.0) -> dict:
    """Flag positions exceeding threshold."""
    alerts = []
    max_pos = {"ticker": None, "weight_pct": 0}

    for ticker, w in weights.items():
        if w["weight_pct"] > max_pos["weight_pct"]:
            max_pos = {"ticker": ticker, "weight_pct": w["weight_pct"]}
        if w["weight_pct"] > threshold:
            alerts.append({
                "ticker": ticker,
                "weight_pct": w["weight_pct"],
                "alert": f"Position exceeds {threshold}% threshold",
            })

    return {
        "alerts": alerts,
        "max_position": max_pos,
        "position_count": len(weights),
    }


def _compute_correlation_matrix(tickers: list[str], period: str = "6mo") -> dict:
    """Correlation matrix from daily returns. Limited to top 10 tickers."""
    tickers = tickers[:10]  # Cap for performance
    if len(tickers) < 2:
        return {"high_pairs": [], "avg_correlation": 0}

    try:
        period_days = {"3mo": 66, "6mo": 130, "1y": 252}.get(period, 130)

        closes_dict = {}
        for t in tickers:
            history = fetch_history_days(t, days=period_days)
            if history:
                closes_dict[t] = pd.Series(
                    [h["close"] for h in history],
                    index=pd.to_datetime([h["date"] for h in history]),
                )

        if len(closes_dict) < 2:
            return {"high_pairs": [], "avg_correlation": 0}

        prices = pd.DataFrame(closes_dict).dropna()
        returns = prices.pct_change().dropna()
        corr = returns.corr()

        # Find highly correlated pairs
        high_pairs = []
        seen = set()
        for i, t1 in enumerate(corr.columns):
            for j, t2 in enumerate(corr.columns):
                if i >= j:
                    continue
                pair_key = tuple(sorted([t1, t2]))
                if pair_key in seen:
                    continue
                seen.add(pair_key)
                c = float(corr.loc[t1, t2])
                if abs(c) > 0.8:
                    high_pairs.append({
                        "pair": [t1, t2],
                        "correlation": round(c, 3),
                        "risk": "High co-movement — limited diversification benefit",
                    })

        # Average correlation (excluding diagonal)
        mask = np.triu(np.ones(corr.shape), k=1).astype(bool)
        avg = float(corr.values[mask].mean()) if mask.any() else 0

        return {"high_pairs": high_pairs, "avg_correlation": round(avg, 3)}

    except Exception:
        return {"high_pairs": [], "avg_correlation": 0}


def _compute_drawdown_analysis(tickers: list[str], quotes: dict) -> dict:
    """Drawdown from 52-week high per holding."""
    drawdowns = {}
    worst = {"ticker": None, "drawdown_pct": 0}

    for t in tickers:
        q = quotes.get(t, {})
        high = q.get("fifty_two_week_high")
        current = q.get("current_price")
        if high and current and high > 0:
            dd = round(((current - high) / high) * 100, 2)
            drawdowns[t] = {"drawdown_pct": dd, "from_high": high, "current": current}
            if dd < worst.get("drawdown_pct", 0):
                worst = {"ticker": t, "drawdown_pct": dd}
        else:
            drawdowns[t] = {"drawdown_pct": None, "from_high": high, "current": current}

    return {**drawdowns, "_worst": worst}


def _compute_stop_loss_alerts(
    tickers: list[str], quotes: dict, default_stop_pct: float = 0.15
) -> list[dict]:
    """Check positions vs trailing stop-loss level (15% from 52wk high)."""
    alerts = []
    for t in tickers:
        q = quotes.get(t, {})
        high = q.get("fifty_two_week_high")
        current = q.get("current_price")
        if not high or not current:
            continue
        stop_level = round(high * (1 - default_stop_pct), 2)
        if current < stop_level:
            alerts.append({
                "ticker": t,
                "current": current,
                "stop_level": stop_level,
                "from_high": high,
                "status": "BELOW_STOP_LOSS",
            })
        elif current < stop_level * 1.05:
            alerts.append({
                "ticker": t,
                "current": current,
                "stop_level": stop_level,
                "from_high": high,
                "status": "NEAR_STOP_LOSS",
            })
    return alerts
